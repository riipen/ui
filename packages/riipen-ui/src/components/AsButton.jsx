import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";
import _JSXStyle from "styled-jsx/style";

const AsButton = ({
  ariaLabel,
  classes,
  children,
  clearButtonStyle,
  type,
  ...other
}) => (
  <>
    <button
      aria-label={ariaLabel}
      className={clsx([
        "button",
        classes,
        clearButtonStyle ? "clearButtonStyle" : null
      ])}
      type={type}
      {...other}
    >
      {children}
    </button>
    <style jsx>{`
      .clearButtonStyle {
        background-color: transparent;
        border: none;
        text-decoration: none;
      }

      .button {
        cursor: pointer;
      }

      .button:focus {
        outline: 5px auto -webkit-focus-ring-color;
      }
    `}</style>
  </>
);

AsButton.propTypes = {
  // external

  /**
   * The aria-label to provide to the button
   */
  ariaLabel: PropTypes.string.isRequired,

  /**
   * The children to render within the button
   */
  children: PropTypes.node.isRequired,

  /**
   * List of additional classes to apply to this component.
   */
  classes: PropTypes.array,

  /**
   * Whether traditional button styles (border, background) should be cleared
   */
  clearButtonStyle: PropTypes.bool,

  /**
   * Whether traditional button styles (border, background) should be cleared
   */
  type: PropTypes.oneOf(["button", "submit", "reset"])
};

AsButton.defaultProps = {
  classes: [],
  clearButtonStyle: true,
  type: "button"
};

AsButton.displayName = "AsButton";

export default AsButton;
