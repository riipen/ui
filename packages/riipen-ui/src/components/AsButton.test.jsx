import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import AsButton from "./AsButton";

describe("AsButton component", () => {
  const child = <div>A child</div>;
  it("renders without errors", () => {
    let error;

    try {
      mount(<AsButton ariaLabel="click to be cool">{child}</AsButton>);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(
      <AsButton ariaLabel="click to be cool">{child}</AsButton>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const wrapper = mount(
        <AsButton ariaLabel="click to be cool">{child}</AsButton>
      );

      expect(wrapper.find("AsButton").props().classes).toEqual([]);
      expect(wrapper.find("AsButton").props().clearButtonStyle).toEqual(true);
      expect(wrapper.find("AsButton").props().type).toEqual("button");
    });
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const wrapper = mount(
        <AsButton ariaLabel="click to be cool">{child}</AsButton>
      );

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classVariant = ["newClass"];

      const wrapper = mount(
        <AsButton ariaLabel="click to be cool" classes={classVariant}>
          {child}
        </AsButton>
      );

      expect(
        wrapper
          .find("AsButton")
          .childAt(0)
          .hasClass(classVariant[0])
      ).toEqual(true);
    });
  });

  describe("clearButtonStyle prop", () => {
    it("Does not apply clearButtonStyle class when clearButtonStyle prop is false", () => {
      const clearButtonStyle = false;

      const wrapper = mount(
        <AsButton
          ariaLabel="click to be cool"
          clearButtonStyle={clearButtonStyle}
        >
          {child}
        </AsButton>
      );

      expect(wrapper.find("AsButton").props().clearButtonStyle).toEqual(
        clearButtonStyle
      );
      expect(
        wrapper
          .find("AsButton")
          .find("button")
          .hasClass("clearButtonStyle")
      ).toEqual(false);
    });

    it("Applies clearButtonStyle class when clearButtonStyle prop is true", () => {
      const wrapper = mount(
        <AsButton ariaLabel="click to be cool">{child}</AsButton>
      );

      expect(wrapper.find("AsButton").props().clearButtonStyle).toEqual(true);
      expect(
        wrapper
          .find("AsButton")
          .find("button")
          .hasClass("clearButtonStyle")
      ).toEqual(true);
    });
  });

  describe("type prop", () => {
    it("Applies type to button", () => {
      const type = "submit";

      const wrapper = mount(
        <AsButton ariaLabel="click to be cool" type={type}>
          {child}
        </AsButton>
      );

      expect(wrapper.find("AsButton").props().type).toEqual(type);
    });
  });

  describe("ariaLabel prop", () => {
    it("Applies ariaLabel to button", () => {
      const ariaLabel = "cool";

      const wrapper = mount(<AsButton ariaLabel={ariaLabel}>{child}</AsButton>);

      expect(wrapper.find("AsButton").props().ariaLabel).toEqual(ariaLabel);
      expect(wrapper.find("button[aria-label='cool']")).toHaveLength(1);
    });
  });
});
