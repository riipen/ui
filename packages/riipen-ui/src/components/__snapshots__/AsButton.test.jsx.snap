// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsButton component renders correct snapshot 1`] = `
<AsButton
  ariaLabel="click to be cool"
  classes={Array []}
  clearButtonStyle={true}
  type="button"
>
  <button
    aria-label="click to be cool"
    className="jsx-3335534469 button clearButtonStyle"
    type="button"
  >
    <div>
      A child
    </div>
  </button>
  <JSXStyle
    id="3335534469"
  />
</AsButton>
`;
